p2panda-net/src/sync/accept.rs

```diff
@@ -0,0 +1,243 @@
+/// Protocol Flow" required for the engine to work efficiently. We're expecting the following
+/// messages from this "acceptor" flow:
+///
+/// 1. `SyncStart`: The sync session just began, we don't know the topic yet.
+/// 2. `SyncHandshakeSuccess`: We've successfully completed the I. "Handshake" phase, as we've
+///    received the topic from the initiator.
+/// 3. `SyncMessage` (optional): The actual data we've exchanged with the other peer, this message
+///    can occur never or multiple times, depending on how much data was sent.
+/// 4. `SyncDone` We've successfully finished this session.
+///
+/// In case of a detected failure (either through an critical error on our end or an unexpected
+/// behaviour from the remote peer), the acceptor will send an `SyncFailed` message instead of the
+/// `SyncDone`.
+///
+/// Errors can be roughly categorized by:
+///
+/// 1. Critical system failures (bug in p2panda code or sync implementation, sync implementation
+///    did not follow "2. Phase Flow" requirements, lack of system resources, etc.)
+/// 2. Unexpected Behaviour (remote peer abruptly disconnected, error which got correctly handled
+///    in sync implementation, etc.)
+pub async fn accept_sync<T, S, R>(
+    mut send: &mut S,
+    mut recv: &mut R,
+    peer: PublicKey,
+    sync_protocol: Arc<dyn for<'a> SyncProtocol<'a, T> + 'static>,
+    engine_actor_tx: mpsc::Sender<ToEngineActor<T>>,
+) -> Result<(), SyncError>
+where
+    T: Topic + TopicId + 'static,
+    S: AsyncWrite + Send + Unpin,
+    R: AsyncRead + Send + Unpin,
+{
+    debug!("accept sync session with peer {}", peer);
+
+    engine_actor_tx
+        .send(ToEngineActor::SyncStart { topic: None, peer })
+        .await
+        .map_err(|err| {
+            SyncError::Critical(format!("engine_actor_tx failed sending sync start: {err}"))
+        })?;
+
+    // Set up a channel for receiving messages from the sync session.
+    let (tx, mut rx) = mpsc::channel::<FromSync<T>>(128);
+    let mut sink = PollSender::new(tx).sink_map_err(|e| SyncError::Critical(e.to_string()));
+
+    // Set up a channel for sending over errors to the "glue" task which occurred during sync.
+    let (sync_error_tx, mut sync_error_rx) = oneshot::channel::<SyncError>();
+
+    // Spawn a "glue" task which represents the layer between the sync session and the engine.
+    //
+    // It picks up any messages from the sync session, making sure that the "Two-Phase Sync Flow"
+    // is followed (I. "Handshake" Phase & II. "Data Sync" Phase), and informs the engine
+    // accordingly.
+    //
+    // If the task detects any invalid behaviour from the sync flow it fails critically, indicating
+    // that the sync protocol implementation does not behave correctly and is not compatible with
+    // the engine.
+    //
+    // Additionally, the task forwards any synced application data straight to the engine.
+    let glue_task_handle: JoinHandle<Result<(), SyncError>> = tokio::spawn(async move {
```

Comment:
Really great well documented code here, we've come a long way with understanding these sync state transitions. Feels like it would even benefit from having it's own struct (`SyncDriver` or something) to make the state transitions even more explicit. Need to think about it more, but it could even be something we expose from the `p2panda-sync` crate :thinking: not anything for now, but will keep it in mind.